### PR TITLE
fix(android/engine): Remove usage of WRITE_EXTERNAL_STORAGE permission

### LINKED ISSUE #4120: bug(android/app): Android request for storage permission hidden by "Getting Started"
Starting in Android M, Android requires apps to that use external storage (SD card) to request for storage permission.
The Keyman app uses this for accessing downloaded kmp packages.

@makara noted in user testing the system dialog is hidden by the "Getting Started" menu. On a fresh install of the app, a user could go through the "Getting Started" steps and install a keyboard from keyman.com before seeing and accepting the system dialog.

We should display "Getting Started" after the user deals with the system dialog.

### PR DESCRIPTION
**Pull request by darcywong00**
Fixes #4120 and reverts #2946

Our current use of DownloadManager with  [setDestinationUri](https://developer.android.com/reference/android/app/DownloadManager.Request#setDestinationUri(android.net.Uri)) meant:
> Must be a file URI to a path on external storage, and the calling application must have the WRITE_EXTERNAL_STORAGE permission.

This mean on app start, we'd need to request WRITE_EXTERNAL_STORAGE permission, and the dialog would get hidden by the "Get Started" menu.

This PR removes the usage of setDestinationUri, which means we can remove the request for WRITE_EXTERNAL_STORAGE permission.

## Additional changes
* Refactor the conversion of the downloaded file from `content://` and `file://` schemes to `DownloadFileUtils`.
* DownloadFileUtils.cacheDownloadFile() also copies the downloaded file to local cache so other cloud callbacks can easily access it (`destinationFile`).

## User Testing
Scenarios to verify the DownloadManager still works

### Get keyboard update
1. I used a custom build that had an older version of sil_euro_latin.kmp (version 1.8.1).
2. Load this version of Keyman for Android
3. Close the "Get Started" menu and allow the app time to check for keyboard updates
4. When a notification for an update to sil_euro_latin appears, download it
5. The keyboard update will proceed in the background. Eventually the keyboard downloads and updates.

For the rest of these tests, use the CI test build

### Install from keyman.com
Downloading a keyboard package doesn't use DownloadManager, but getting an associated dictionary does.
1. Load Keyman for Android and close the "Get Started" menu
2. Go to "Settings" --> Install Keyboard or Dictionary --> Install from keyman.com
3. Search for **gff_amharic** keyboard and install
4. Verify the keyboard package downloads and can be installed
5. Eventually, the associated lexical model will be downloaded and installed too.

### Install from local file
1. From Chrome, download another keyboard package and save/copy it to the SD card/Downloads folder (If there's no Downloads folder on SD card, make it).
2. Load Keyman for Android and close the "Get Started" menu
3. Go to "Settings" --> Install keyboard or Dictionary --> Install from local file
4. Browse to SD card / Downloads  to the saved kmp file and select it
5. Verify Keyman for Android shows a system dialog requesting access to the SD card
6. Accept the permission and continue with the keyboard package installation

### CODE REVIEW

**Comment by MattGyverLee** [comment]
This is great. I was wondering why I was alerted about new permissions when I moved to embedding KMEA13.

**Comment by darcywong00** [comment]
>  I was wondering why I was alerted about new permissions when I moved to embedding KMEA13.

The `WRITE_EXTERNAL_STORAGE` permission was being used in KMEA 14 alpha. KMEA 13 is only using `READ_EXTERNAL_STORAGE` (and probably quietly throwing storage exceptions).

**Comment by mcdurdin** [comment]
Well done! I am happy to see this PR. I will review when back at my desk.

**Review by mcdurdin — CHANGES_REQUESTED** [review]
This looks really good. I like the refactor. I've got some minor change requests, and I realise some of them come from the existing code that was simply shifted around, but I think that it's important to leave the code in a better state than it was anyway ;-)

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+          cachedFile.delete();
+        }
+
+        FileUtils.copy(inputFile, new FileOutputStream(cachedFile));
+      }
+    } catch (Exception e) {
+      String message = "Access denied to " + filename +
+        ".\nCheck Android Settings --> Apps --> Keyman to grant storage permissions";
+      KMLog.LogException(TAG, "Unable to copy " + filename + " to app cache ", e);
+      Toast.makeText(context, message, Toast.LENGTH_LONG).show();
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java]: Done. I had the two callers (`MainActivity.useLocalKMP()` and `CloudApiTypes.cacheAndOpenDestinationFile()`) check and handle the Toast notifications.

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ */
+public final class DownloadFileUtils {
+  private static final String TAG = "DownloadFileUtils";
+
+  /**
+   * Small class for returning information about a file downloaded via DownloadManager.
+   */
+  public static final class Info {
+    private final boolean isKMP;
+    private final String filename;    // Internal filename (may be a document number)
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java]: That was a stale comment. I've updated it

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      }
+    } catch (Exception e) {
+      String message = "Access denied to " + filename +
+        ".\nCheck Android Settings --> Apps --> Keyman to grant storage permissions";
+      KMLog.LogException(TAG, "Unable to copy " + filename + " to app cache ", e);
+      Toast.makeText(context, message, Toast.LENGTH_LONG).show();
+    }
+
+    Info info = new Info(isKMP, filename, cachedFile);
+    return info;
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java:99]: The caller in MainActivity will return if this is invalid

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudApiTypes.java [hunk 4]
```diff
@@ -140,8 +142,12 @@ public long getDownloadId() {
       return downloadId;
     }
 
-    public File getDestinationFile() {
-      return destinationFile;
+    public File getDestinationFile(Context context) {
+      DownloadManager downloadManager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
+      Uri data = downloadManager.getUriForDownloadedFile(downloadId);
+      DownloadFileUtils.Info info = DownloadFileUtils.cacheDownloadFile(context, data);
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/cloud/CloudApiTypes.java:155]: Good idea. I've renamed to `cacheAndOpenDestinationFile()`

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java [hunk 5]
```diff
@@ -820,57 +811,20 @@ public static void useLocalKMP(Context context, Uri data) {
...
-      Toast.makeText(context, message, Toast.LENGTH_LONG).show();
-      return;
+    DownloadFileUtils.Info info = DownloadFileUtils.cacheDownloadFile(context, data);
+    boolean isKMP = info.isKMP();
+    String filename = info.getFilename();
+    File cacheKMPFile = info.getFile();
+
+    if (!isKMP) {
+      String noKeyboardsInstalledMessage = " is not a valid Keyman package file.\n" +
+        "No keyboards/dictionaries were installed.";
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/MainActivity.java]: Added to localizable string R.string.not_valid_package_file

File: android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        cachedFile = new File(context.getCacheDir().toString(), filename);
+        if (cachedFile.exists()) {
+          cachedFile.delete();
+        }
+
+        FileUtils.copy(inputFile, new FileOutputStream(cachedFile));
+      }
+    } catch (Exception e) {
+      String message = "Access denied to " + filename +
+        ".\nCheck Android Settings --> Apps --> Keyman to grant storage permissions";
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMEA/app/src/main/java/com/tavultesoft/kmea/util/DownloadFileUtils.java:94]: The callers now check if filename is null and give a Toast and return.

**Review by mcdurdin — APPROVED** [review]
LGTM.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.220-beta](https://keyman.com/downloads/releases/beta/14.0.220)

**Comment by MakaraSok** [comment]
@darcywong00 Tested against this build (v14.0.220-beta) on emulator Mixel 2 API 29.

**One issue is found**

At this step of "Install from local file":

> 5. Verify Keyman for Android shows a system dialog requesting access to the SD card

No dialog requesting access to the SD card was shown to me. The keyboard got installed without any blockage.

**Comment by darcywong00** [comment]
> No dialog requesting access to the SD card was shown to me. The keyboard got installed without any blockage.

Ah, well I'm not concerned if some Android devices don't ask for the permission. I know older devices don't bother asking.

**Comment by MakaraSok** [comment]
Fair enough then :)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.9-alpha](https://keyman.com/downloads/releases/alpha/15.0.9)

**Pull request merged**